src/main/java/com/uber/cadence/internal/shadowing/ReplayWorkflowActivityImpl.java

```diff
@@ -0,0 +1,225 @@
+import com.uber.m3.tally.Stopwatch;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ReplayWorkflowActivityImpl implements ReplayWorkflowActivity {
+
+  private static final Logger log = LoggerFactory.getLogger(ReplayWorkflowActivityImpl.class);
+
+  private final IWorkflowService serviceClient;
+  private final Scope metricsScope;
+  private final Class<?> workflowClass;
+  private final Class<?>[] workflowClasses;
+
+  public ReplayWorkflowActivityImpl(
+      IWorkflowService serviceClient,
+      Scope metricsScope,
+      Class<?> workflowClass,
+      Class<?>... workflowClasses) {
+    this.serviceClient = Objects.requireNonNull(serviceClient);
+    this.metricsScope = Objects.requireNonNull(metricsScope);
+    this.workflowClass = Objects.requireNonNull(workflowClass);
+    this.workflowClasses = workflowClasses;
+  }
+
+  @Override
+  public ReplayWorkflowActivityResult replay(ReplayWorkflowActivityParams request) {
+    if (request == null) {
+      throw new NullPointerException("Replay activity request is null.");
+    }
+
+    String domain = request.getDomain();
+    int successCount = 0;
+    int failedCount = 0;
+    int skippedCount = 0;
+    int replayIndex = 0;
+    List<WorkflowExecution> executions = request.getExecutions();
+
+    // Retrieve process from heartbeat
+    Optional<HeartbeatDetail> heartbeatDetail = Activity.getHeartbeatDetails(HeartbeatDetail.class);
+    if (heartbeatDetail.isPresent()) {
+      ReplayWorkflowActivityResult heartbeatResult = heartbeatDetail.get().getReplayResult();
+      successCount = heartbeatResult.getSucceeded();
+      failedCount = heartbeatResult.getFailed();
+      skippedCount = heartbeatResult.getSkipped();
+      replayIndex = heartbeatDetail.get().getReplayExecutionIndex() + 1;
+    }
+
+    for (; replayIndex < executions.size(); replayIndex++) {
+      WorkflowExecution execution = executions.get(replayIndex);
+      ReplayWorkflowActivityResult oneReplayResult = replayHelper(request.getDomain(), execution);
+      successCount += oneReplayResult.getSucceeded();
+      failedCount += oneReplayResult.getFailed();
+      skippedCount += oneReplayResult.getSkipped();
+      ReplayWorkflowActivityResult heartbeatResult =
+          new ReplayWorkflowActivityResult()
+              .setSucceeded(successCount)
+              .setFailed(failedCount)
+              .setSkipped(skippedCount);
+      Activity.heartbeat(new HeartbeatDetail(heartbeatResult, replayIndex));
+    }
+    return new ReplayWorkflowActivityResult()
+        .setSucceeded(successCount)
+        .setFailed(failedCount)
+        .setSkipped(skippedCount);
+  }
+
+  public ReplayWorkflowActivityResult replayHelper(String domain, WorkflowExecution execution) {
+    WorkflowExecutionHistory workflowHistory;
+    try {
+      workflowHistory = getFullHistory(domain, execution);
+    } catch (Exception e) {
```

Comment:
for workflow type not defined error, shall we failed the workflow instead of skipping it? I think it usually means there's a something wrong with workflow registration and require a worker deployment to fix it. 

In go client, I returned a non-retryable error defined in thrift for the activity so that the workflow will fail. What's your opinion?

Reply:
I am thinking of it as a replay failure and emit some replay metrics. The use case in my mind is, for example, there are 2 wf types. one is registered and one doesn't. We should replay for those 2 wf types and emit failure count on the unregister ones. What do you think?